keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,8 +1069,191 @@ the remaining requests in that queue start getting faster service.  In
+our estimations here are very rough anyway that seems a reasonable
+simplification that makes dispatching the budget a bit simpler.
+
+#### Determining the width
+
+While one can imagine arbitrarily sophisticated algorithms for it (including
+exposing defining the width of requests via FlowSchema API), we want to start
+with something relatively simple to first get operational experience with it
+before investing into sophisticated algorithms or exposing a knob to users.
+
+In order to determine the function that will be approximating the `width` of
+a request, we should first estimate how expensive a particular request is.
+And we need to think about both dimensions that we're trying to protect from
+overloading (CPU and RAM) and how many concurrency units a request can actually
+consume.
+
+Let's start with CPU.  The total cost of processing a LIST request should be
+proportional to the number of processed objects.  However, given that in
+practice processing a single request isn't parallelized (and the fact that
+we generally scale the number of total concurrency units linearly with amount
+of available resources), a single request should consume no more than A
+concurrency units.  Fortunately that all compiles together because the
+`processing latency` of the LIST request is actually proportional to the
+number of processed objects, so the cost of the request (defined above as
+`<width> x <processing latency>` really is proportaional to the number of
+processed objects as expected.
+
+For RAM the situation is actually different.  In order to process a LIST
+request we actually store all objects that we process in memory. Given that
+memory is uncompressable resource, we effectively need to reserve all that
+memory for the whole time of processing that request.  That suggests that
+the `width` for the request from the RAM perspective should be proportional
+to the number of processed items.
+
+So what we get is that:
+```
+  width_cpu(N) = min(A, B * N)
+  width_ram(N) = D * N
+```
+where N is the number of items a given LIST request is processing.
+
+The question is how to combine them to a single number.  While the main goal
+is to stay on the safe side and protect from the overload, we also want to
+maxiumize the utilization of the available concurrency units.
+Fortunately, when we normalize CPU and RAM to percentage of available capacity,
+it appears that almost all requests are much more cpu-intensive.  Assuming
+4GB:1CPU ratio and 10kB average object and the fact that processing larger
+number of objects can utilize exactly 1 core, that means that we need to
+process 400.000 objects to make the memory cost higher.
+This means, that we can afford the potential minor efficiency that extremely
+large requests would cause and just approximate it by protecting every resource
+independently, which translates to the following function:
+```
+  width(n) = max(min(A, B * N), D * N)
```

Comment:
You can not faithfully crunch a two-dimensional bin-packing problem, with different shape blocks (which appears both because of the mix of LIST and non-LIST requests and because the LIST requests do not always have the same proportion between CPU and MEM), into one dimension.

Reply:
I'm not solving bin-packing problem. I'm solving "overload protection" problem.

The main goal is protect the server from overload, not to exactly use all the available capacity. I much more prefer to waste some percentage of capacity, than to risk overloading.

So this isn't binpacking problem - maxking the max from dimensions is definitely inefficient from the binpacking POV, but it solves the aspect of overloading, which is the main goal.